flight/PiOS/STM32F4xx/pios_dmashot.c

```diff
@@ -0,0 +1,591 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, see <http://www.gnu.org/licenses/>
+ *
+ * Additional note on redistribution: The copyright and license notices above
+ * must be maintained in each individual source file that is a derivative work
+ * of this source file; otherwise redistribution is prohibited.
+ */
+
+#include <pios.h>
+#include "pios_dmashot.h"
+
+#define MAX_TIMERS                              8
+
+// This is to do half-word writes where appropriate. TIM2 and TIM5 are 32-bit, the rest
+// are 16-bit. Can do full word writes to 16bit regardless, but it doubles DMA buffer size
+// unnecessarily.
+union dma_buffer {
+	void *ptr;
+	uint32_t *fw;
+	uint16_t *hw;
+};
+
+// Internal structure for timer and DShot configuration.
+struct servo_timer {
+	uint32_t sysclock;                                                              // Flava Flav! Timer's clock frequency.
+
+	const struct pios_dmashot_timer_cfg *dma;                                       // DMA config
+	const struct pios_tim_channel *servo_channels[4];                               // DMAShot configured channels
+
+	// Tracking used TIM channels to allocate least memory.
+	uint8_t low_channel;                                                            // Lowest registered TIM channel
+	uint8_t high_channel;                                                           // Duh
+
+	uint32_t dshot_freq;                                                            // Stores the desired frequency for
+	                                                                                // timer initialization
+	uint16_t duty_cycle_0;                                                          // Calculated duty cycle for 0-bit
+	uint16_t duty_cycle_1;                                                          // And for 1-bit
+
+	union dma_buffer buffer;                                                        // DMA buffer
+};
+
+// DShot signal is 16-bit. Use a pause before and after to delimit signal and quell the timer CC
+// when the message ends.
+#define DMASHOT_MESSAGE_WIDTH                   16
+#define DMASHOT_MESSAGE_PAUSE                   1
+#define DMASHOT_STM32_BUFFER                    (DMASHOT_MESSAGE_PAUSE + DMASHOT_MESSAGE_WIDTH + DMASHOT_MESSAGE_PAUSE)
+
+// Those values are in percent. These represent the Betaflight/BLHeli_s timing,
+// instead of the spec one. Works with KISS RE 24A though.
+#define DSHOT_DUTY_CYCLE_0                      36
+#define DSHOT_DUTY_CYCLE_1                      74
+
+#define TIMC_TO_INDEX(c)                        ((c)>>2)
+
+const struct pios_dmashot_cfg *dmashot_cfg;
+struct servo_timer **servo_timers;
+
+static inline int PIOS_DMAShot_GetFIFOCadence(int num_chan)
```

Comment:
They need to be chosen from the table to be a legal transfer for the **memory burst size** and element size.  But the memory burst size doesn't need to match the peripheral operation size, and the burst thing that the timer does is different from a DMA controller burst anyways.

Basically, every time the FIFO isn't filled, the DMA tries to get control of the memory bus, and do a sequence of N transfers in order to fill the fifo up to the desired level.  Because turning around the memory bus takes time and sequential transfers are faster, this uses the memory bandwidth more efficiently.  Some combinations are not legal (e.g. ones that risk overfilling the FIFO)

I believe at least-- this is my interpretation.